### ISSUE TITLE
Proposing another implementation of lisp-modules

### ISSUE THREAD

**Issue by Uthar** [NixOS/nixpkgs#155851]
Hi!

I made a small [nix library](https://github.com/uthar/nix-cl) (`nix-cl`), for working with ASDF and Quicklisp, which has been working great for me recently. The code is still a mess but I want to show it to you and see if it sparks interest. 

Basically I think it could replace the current `lisp-modules` and provide a number of benefits.

If you'd like to test it out you should use `direnv` or source `.envrc` to get the examples working. (Have to set NIX_PATH). 
(Also worth noting, I think that `nixpkgs` doesn't provide Clasp but it worked with my own version.)

I had these problems with `lisp-modules` - please correct me if I'm wrong here:

- Had to read the source to find how to run a lisp wrapper (`common-lisp.sh`) - no documentation
- It leaks FASL's into `$HOME` #142604
- Hard to work with Quicklisp (have to build everything ahead of time, which is slow): #150941
- CFFI doesn't work on ABCL (the one library that I know of which requires a `.jar` in the `CLASSPATH`)
- Can't have multiple lisps in one `nix-shell` because a lisp-with-packages is achieved by putting the libraries themselves in `buildInputs` and running the generic `common-lisp.sh`
- Doesn't support ASDF systems where `.asd` is **not** in the root directory (e.g. [jzon](https://github.com/Zulu-Inuoe/jzon))

Which are improved in `nix-cl`:
- Has some documentation on how to run a `${lisp}WithPackages`
- `CL_SOURCE_REGISTRY`, `ASDF_OUTPUT_TRANSLATIONS` are set in a granular way, per package
- It  generates nix expressions for **everything** in quicklisp, which works well because of Nix's laziness. One can then just pick what they want from that.
- Can declare Java libraries for ABCL
- `abclWithPackages`/`cclWithPackages`/`eclWithPackages`/`sbclWithPackages` etc. makes it possible to have multiple lisps in one `nix-shell`
- Works for `asd`'s not in root by using "//" suffix to CL_SOURCE_REGISTRY (cool feature of ASDF - [manual](https://asdf.common-lisp.dev/asdf/Shell_002dfriendly-syntax-for-configuration.html#Shell_002dfriendly-syntax-for-configuration))
- Easy to package things not in quicklisp. Minimal package is `pname`,`version`,`src`.

There are issues with it still where I haven't had the need to fix them:
- CLASSPATH handling assumes `/share/java/*.jar` (from `fetchMavenArtifact`)
- Haven't handled the case where the manually defined packages could use the quicklisp-imported ones as dependencies (worked around by just copying the code to manual packages)
- Java jars are included unnecessarily in non-java lisp wrappers
- Weird prefix for libs beginning with a number (e.g. `_3d-vectors`)

@7c6f434c

**Comment by 7c6f434c** [comment]
Copying manually the generated definitions for anything that depends on them sounds pretty bad (in the context of Nyxt being packaged like a non-Quicklisp package that depends on large amounts of whatever)…

Is compilation supposed to happen for all the stuff in withPackages at once?

**Comment by Uthar** [comment]
yeah it's all pre-compiled at once

when it comes to compilatioin, there was another problems that I had:
- cl-unicode generates files during compilation, which failed when ASDF output is set to storeDir. Workaround was to compile first into pwd, then use $out of that as an $src of the next build

**Comment by Uthar** [comment]
I haven't had problems with copying generated definitions. They are almost the same as the manually specified ones, with the exception that when building a system like `cl-async-util` which definition of is in `cl-async.asd`, a file `cl-async-util.asd` is created first. `asd`'s other than the name of the system being built are removed to prevent runtime references to uncompiled sytems, so without such a hack systems like this `cl-async-util`  were unloadable. 

Now that  I think about it it would have to be the same for the manual package, but I haven't had such a package yet. So in that case the ql imported things is just a way to save a lot of typing.

**Comment by 7c6f434c** [comment]
I have nothing against some special-case workarounds; I somewhat like the idea of binary cache compatibility, though.

About copying: it's trivial, but it needs some integration with mass updates for manual packages with tons of dependencies.

**Comment by Uthar** [comment]
Hmmm I understand, that is problematic

interaction between the imported and manual packages is kind of tricky

Maybe I could load the quicklisp packages without rewriting their dependency trees at first, then load the manual packages and provide quicklisp imports as additional libraries. Just to break the current inter-dependency. Then go through the resulting thing and do whatever such as prefer manual libs to ql. Then build that derivation

**Comment by 7c6f434c** [comment]
Maybe a generator could do all the copying before use, though.

**Comment by Uthar** [comment]
Are you thinking about a separate script that updates the file containing the manual packages?  That could work . Or even a separate file, "ql-deps.nix" or something to keep the manual file clean

**Comment by 7c6f434c** [comment]
Basically how the current Lisp infrastructure just generates the Quicklisp data once as a nix file, yes.

**Comment by Uthar** [comment]
> the case where the manually defined packages could use the quicklisp-imported ones as dependencies

Added that in https://github.com/Uthar/nix-cl/commit/a0eb90a9be566cda33c05d1891c063b4370e4335

It was pretty productive because I could remove most of my package code and just reuse the QL generated things. So thanks a lot for this suggestion @7c6f434c! 

For example, to pin the version of `dexador` which has these dependencies (`lispLibs`):
```
alexandria babel bordeaux-threads chipz chunga cl+ssl
cl-base64 cl-cookie cl-ppcre fast-http fast-io quri
trivial-features trivial-gray-streams trivial-mimes uiop usocket
```
this code in `packages.nix` works:
```
dexador = build-asdf-system {
  inherit (ql.dexador) pname lispLibs;
  version = "20220115";
  src = builtins.fetchTarball {
    url = "https://github.com/fukamachi/dexador/archive/03aaadf18b3cdcfa625627587951fa956bdf490f.tar.gz";
    sha256 = "0hwmzw9l2iqg4v0x59h4amr8bxmxdpc9nd8a3k6xji8g7v25idxw";
  };
};
```
Later on, it will simply pick up updates to its dependencies from the quicklisp generated file. At the same time, anything that depends on `dexador`, either QL-imported or declared manually, will now use *this* version of it.

Though now it's redundant with the `extras` mechanism in `ql.nix`. Meaning *some* customizations can be done in either `ql.nix` or `packages.nix`. Hmmm not sure what to do about that. Myself I would keep non-QL packages in `packages.nix`, and trivial fixes to QL packages (addded native library etc.) in `ql.nix`.

**Comment by 7c6f434c** [comment]
>Though now it's redundant with the `extras` mechanism in `ql.nix`. Meaning *some* customizations can be done in either `ql.nix` or `packages.nix`. Hmmm not sure what to do about that. Myself I would keep non-QL packages in `packages.nix`, and trivial fixes to QL packages (addded native library etc.) in `ql.nix`.

Which is also more or less what is done now (which is surely a good thing)

Do I understand correctly that you trust Quicklisp dependency data?

(Hmm, now if only I understood how to combine your code with binary-cache-friendlyness/precompilation, it could indeed be attractive to just replace lispPackages with that)

**Comment by Uthar** [comment]
I do directly read the quicklisp dependency data, then do some merging on it in the importer to capture dependencies of dependencies etc. 
 
Most stuff that I tried building this way just worked - in case of problems a common one was a dependency on a "slashy" system. I ignore those in the importer because to it, a unit of granularity is the system, not the `.asd` file (that's why these files are created as necessary) - but "slashy" systems belong to the "root-name-part.asd" file. At the time I didn't know how to handle that, but maybe there is a way.

One of the cool things that I'm thinking about would be the ability to run Nix builds of the entire Quicklisp on Nix orgs' Hydra servers - probably a gazillion bugs would be uncovered. At least for sbcl/ccl where they have pretty fast compilers (?) Not really sure how that would look from resoures/political side.

Could you expand some more about the precompilation concept? If the concern is recompiling all libs for each wrapper, then that's not the case - the lisp packages themselves get compiled once (`sbclPackages`), then any lisp wrappers (`sbclWithPackages`) reuse that without recompiling anything (they just put together a wrapper shell script). Sorry, maybe I didn't understand your previous question correctly so the answer could be misleading.

**Comment by Uthar** [comment]
My current workaround for slashy systems is to add them to the `systems` list of the thing that is being depended on. So, `nyxt` depends on `iolib/pathname` then I set `iolibs.systems` to ` [ "iolib" "iolib/pathname" ]`

**Comment by 7c6f434c** [comment]
>I do directly read the quicklisp dependency data, then do some merging on it in the importer to capture dependencies of dependencies etc. 
> 
>Most stuff that I tried building this way just worked - in case of problems a common one was a dependency on a "slashy" system. I ignore those in the importer because to it, a unit of granularity is the system, not the `.asd` file (that's why these files are created as necessary) - but "slashy" systems belong to the "root-name-part.asd" file. At the time I didn't know how to handle that, but maybe there is a way.

There have been Quicklisp releases in the past with some more interesting problems in dependency data (which is where quicklisp-to-nix takes its motivation).

>One of the cool things that I'm thinking about would be the ability to run Nix builds of the entire Quicklisp on Nix orgs' Hydra servers - probably a gazillion bugs would be uncovered. At least for sbcl/ccl where they have pretty fast compilers (?) Not really sure how that would look from resoures/political side.

Even now CCL doesn't build 100% of lispPackages. SBCL might be doable. Quicklisp is actually somewhat small, so running everything on SBCL will probably end up doable even locally. And as we monotonically grow the set of packages we grab, just maxing it would not be that much of a surprise… 

I don't think we'd uncover any meaningful bugs, because entire Quicklisp is built before release and a package needs to build at least somewhere to not be removed. We would get a somewhat better tool for impact measurement for SBCL upgrades, though.

>Could you expand some more about the precompilation concept? If the concern is recompiling all libs for each wrapper, then that's not the case - the lisp packages themselves get compiled once (`sbclPackages`), then any lisp wrappers (`sbclWithPackages`) reuse that without recompiling anything (they just put together a wrapper shell script). Sorry, maybe I didn't understand your previous question correctly so the answer could be misleading.

Ooh, that's better, I misunderstood you, sorry. I interpreted some of the «all together» as compilation of the packages only once they are all gathered together into an environment.

**Comment by 7c6f434c** [comment]
>My current workaround for slashy systems is to add them to the `systems` list of the thing that is being depended on. So, `nyxt` depends on `iolib/pathname` then I set `iolibs.systems` to ` [ "iolib" "iolib/pathname" ]`

Presumably, if you already grab the tarballs in the generator, you can as well unpack them and obtain the list of such systems, right?

**Comment by Uthar** [comment]
> There have been Quicklisp releases in the past with some more interesting problems in dependency data (which is where quicklisp-to-nix takes its motivation).

That's interesting. I only imported the QL releases 2021-12-30 and 2021-08-07  (that one had a bug with a missing dependency of `clml`). But the newer one removed quite some packages.

> Even now CCL doesn't build 100% of lispPackages. SBCL might be doable. Quicklisp is actually somewhat small, so running everything on SBCL will probably end up doable even locally. And as we monotonically grow the set of packages we grab, just maxing it would not be that much of a surprise…

Hmm you're right. I could actually try building the whole of QL locally using SBCL (Currently that's 4310 packages). Or maybe start with the subset that's in `lispPackages` currently to compare both implementations.



> Presumably, if you already grab the tarballs in the generator, you can as well unpack them and obtain the list of such systems, right?

That could be an option, but Quicklisp's [systems.txt](https://beta.quicklisp.org/dist/quicklisp/2021-12-30/systems.txt) already contains those systems in the third column. I just skipped them for now for lack of a better idea. On the other hand, with 2021-12-30, `nyxt` seems to have a dependency  on `iolib/pathname` that's undeclared in systems.txt. I have to check whether QL info even includes such systems in the dependencies column.

The problem for me was to somehow make slashy systems work with `load-system`. The reason that's problematic is: 
1. There can only be one `foo` system in a unique `foo.asd` file 
2. ASDF looks up `foo/<...>`'s in that same `foo.asd` 

So in effect it "packs" a bunch of systems into a single `.asd` file, which is completely against what `nix-cl` understands.

There is the possiblity to simply always build every slashy subsystem of a system. But that could possibly add unnecessary dependencies in cases the slashy systems won't be used. Because real libraries use the `foo/bar` form, I don't see a possibility to generate some unique names from that either, unless patching ASDF itself. So maybe the "add slashy systems on an as-needed basis" approach is fine after all?

**Comment by Uthar** [comment]
OK, I'll build the systems in `quicklisp-to-nix-systems.txt` using `nix-cl` and post an update.

**Comment by 7c6f434c** [comment]
I think the massive numbers of missed dependencies were quite long ago now.

slashy systems have yet another annoying habit of creating circular dependencies between .asd files, yeah.

**Comment by Uthar** [comment]
Results for `quicklisp-to-nix-systems.txt`:
```
 SBCL: 252/252
  CCL: 249/252
  ECL: 234/252
 ABCL: 210/252
Clasp: 202/252
```

Gonna do the same for things in `pkgs.lispPackages`

**Comment by 7c6f434c** [comment]
From a glance, looks about right.

**Comment by Uthar** [comment]
Status of `lispPackages` for SBCL and CCL:
```
SBCL: 379/384
 CCL: 375/384
```

Things that don't build are the slashy systems:
```
cl-postgres_slash_tests
cxml_slash_test
s-sql_slash_tests
simple-date_slash_postgres-glue
xpath_slash_test
```

And, additionally for CCL, three packages that are explicitly SBCL-only
```
nyxt
stumpwm
woo
```
Plus `varjo`, which has a CCL-specific bug upstream: https://github.com/cbaggers/varjo/pull/243 - fixed but not yet in Quicklisp.

I would propose another solution to the slashy systems problem. On branch `overridable-drv`, I stole `overridePythonAttrs` from `python-packages.nix` and made it into `overrideLispAttrs`. So slashy-systems could now be built in an ad-hoc way.

Here's a `cl-postgres`+`cl-postgres/tests`:
```
sbclPackages.cl-postgres.overrideLispAttrs (old: {
  systems = old.systems ++ [ "cl-postgres/tests" ];
  lispLibs = old.lispLibs ++ [ sbclPackages.fiveam ];
})
```
This could even be used when a certain package depends on some slashy systems, to provide the dependency in an "anonymous" way - without cluttering the original `sbclPackages.cl-postgres`.

Given their "optionality" nature this seems pretty reasonable to me - what do you think?

**Comment by Uthar** [comment]
Got a question about this [line](https://github.com/Uthar/nix-cl/commit/d6b25275e20ee9132cae6b3e4a7575c29e8db890#diff-613b1051f6da9967c6f0a3d28ea9c2ac7a1cd5845510af6945d45bd635ad5c55R249)
This is what I had to do to make nyxt work for me, but not sure if this should be hard-coded like that? Maybe it's just a thing of my particular DE/setup?

**Comment by 7c6f434c** [comment]
>This could even be used when a certain package depends on some slashy systems, to provide the dependency in an "anonymous" way - without cluttering the original `sbclPackages.cl-postgres`.
>
>Given their "optionality" nature this seems pretty reasonable to me - what do you think?

I think there are completely optional (e.g. tests) things like that that seem indeed just skippable, a few places where there is no circular dependencies involved, and I think one or two loops that I somewhat hate where downstream-useful libraries are locked in a circular between-asd dependency. But we don't handle that currently anyway.

**Comment by 7c6f434c** [comment]
Hm, we do not seem to need this now. Does building an older version still need it in your setup?

**Comment by Uthar** [comment]
I get the error with `webkitgtk-2.34.4` (using nixpkgs master) , but only a warning with `webkitgtk-2.32.3` (some old commit, ~20.09) when that variable is not set.

**Comment by 7c6f434c** [comment]
Hm, probably not DE-specific (not sure what their sandbox wants though)

**Comment by Uthar** [comment]
To be precise, this is what happens:
1. webkitgtk-2.34.4
```
** (nyxt:1097805): ERROR **: 15:18:18.273: Unable to configure xdg-desktop-portal access in the WebKit sandbox: GApplication is required.
<WARN> [15:18:18] Warning: Error on GTK thread: The value
  :INVALID-CODE-OBJECT-AT-PC
is not of type
  FUNCTION

** (process:1097805): ERROR (recursed) **: Unable to configure xdg-desktop-portal access in the WebKit sandbox: GApplication is required.fatal error encountered in SBCL pid 1097805 tid 1097819:
SIGABRT received.

   0: fp=0x2 pc=0x7fbbf4f08baa Foreign function gsignal
```

2. webkitgtk-2.32.3

```

** (nyxt:1097629): WARNING **: 15:18:03.123: GApplication is required for xdg-desktop-portal access in the WebKit sandbox. Actions that require xdg-desktop-portal will be broken.

** (nyxt:1097629): WARNING **: 15:18:03.141: GApplication is required for xdg-desktop-portal access in the WebKit sandbox. Actions that require xdg-desktop-portal will be broken.
```

**Comment by 7c6f434c** [comment]
Ah, that might be dependent on having something installed in the profile, possibly via DE in system path…

Re: slashy systems — I think current generation logic defaults to including these systems and blocking the problematic ones. Maybe it's better to keep that approach?

**Comment by Uthar** [comment]
Posting rest of CL implementations for completeness (`lispPackages`):
```
  ECL: 351/384
 ABCL: 322/384
Clasp: 288/384
```

**Comment by Uthar** [comment]
> Re: slashy systems — I think current generation logic defaults to including these systems and blocking the problematic ones. Maybe it's better to keep that approach?

Hmm, Yes that might a better default by providing more things out of the box. I will experiment with that - I have some ideas on how to handle `.asd` files that popped up while trying to build `cl-opengl/es2`. Though with a lot of packages in a wrapper things might get ugly with them. But maybe some error checking could be done on the Nix side to block such scenarios.

**Comment by 7c6f434c** [comment]
>Hmm, Yes that might a better default by providing more things out of the box. I will experiment with that - I have some ideas on how to handle `.asd` files that popped up while trying to build `cl-opengl/es2`. Though with a lot of packages in a wrapper things might get ugly with them. But maybe some error checking could be done on the Nix side to block such scenarios.

Probably on the generator side too…

**Comment by Uthar** [comment]
For now I've been adding stuff on my `slashy` branch:
1, Enabled importing of slashy systems
2. In the generator - When a slashy system depends explicitly on its parent system, move that dependency from `lispLibs` to `systems` - this way they can be built in isolation
3. In the lisp-wrapper-builder - Detected the case when multiple packages in the dependency tree provide the same asd file, and error out on that
4. Attempted to recover from 3. by merging into one parent system

Currently thinking about how to fix the problems:
1. QL systems.txt doesnt include the fact that a system depends on a slashy system. E.g nyxt depends on iolib/os. (which is for now the only practical example I encountered  when 
 building the whole `lispPackages`). These dependencies can be seen with asdf:find-system, though. Hmm I'm feeling the circular dependencies coming with that.

2. Is there a clean way to rewrite a dependency on the parent system that is somewhere deep in the dependency tree when wanting to add a slashy system in a lisp wrapper? 

These last are probably way out there. But eventually I'd like to fix them.

I can build `cl-postgres_slash_tests`, `cxml_slash_test`, `xpath_slash_test` now
But still not  `s-sql_slash_tests` and `simple-date_slash_postgres-glue`

**Comment by Uthar** [comment]
Going off the rails a little bit on the `slashy-recover` branch, with the amount of Nix code...

I think I figured out the 2nd problem about rewriting dependencies.

I modified `lispWithPackages` so that slashy systems get "merged" into their parent system troughout the dependency tree.

As for an example of what I mean,

(`lisp-stat` depends on `data-frame`, which depends on `alexandria+`)

2. 
```
nix-repl> x = sbclWithPackages (ps: [ ps.lisp-stat ])
nix-repl> x.lispLibs                                                                                     
[ «derivation /nix/store/gnh7yj61rmc2lp5ski2qn0skr0fw6inx-lisp-stat-20211209-git.drv» ]

nix-repl> data-frame = builtins.elemAt (builtins.head x.lispLibs).lispLibs 2
«derivation /nix/store/715ckx2m569k9bcngs8xy7277kalncv9-data-frame-20211209-git.drv»

nix-repl> alexandria-plus = (builtins.elemAt data-frame.lispLibs 1) 
«derivation /nix/store/9bnghf94a0lazv25bjkdsmbi20jslkaw-alexandria+-20211230-git.drv»

nix-repl> alexandria-plus.systems 
[ "alexandria+" ]

nix-repl> alexandria-plus.lispLibs
[ «derivation /nix/store/wr6ij2dl3k82brbqa45kbbmr1s64vc2q-alexandria-20211209-git.drv» ]
```
3. 
```
nix-repl> y = sbclWithPackages (ps: [ ps.lisp-stat ps.alexandria_plus_slash_tests ])
nix-repl> y.lispLibs
[ «derivation /nix/store/yqn0lmifmnhxiq3hfb5xap9gvwxmd65m-lisp-stat-20211209-git.drv» «derivation /nix/store/0lqnchxj481lcyw6rh6zrwqf6gmzs3fa-alexandria+-20211230-git.drv» ]

nix-repl> data-frame = builtins.elemAt (builtins.head y.lispLibs).lispLibs 2
«derivation /nix/store/nv6d7cr5nh5mghnc0xxc99xy2vwxjmvx-data-frame-20211209-git.drv»

nix-repl> alexandria-plus = (builtins.elemAt data-frame.lispLibs 1) 
«derivation /nix/store/0lqnchxj481lcyw6rh6zrwqf6gmzs3fa-alexandria+-20211230-git.drv»
(Same as in y.lispLibs!)

nix-repl> alexandria-plus.systems
[ "alexandria+" "alexandria+/tests" ]

nix-repl> alexandria-plus.lispLibs
[ «derivation /nix/store/wr6ij2dl3k82brbqa45kbbmr1s64vc2q-alexandria-20211209-git.drv» «derivation /nix/store/jn9c9rzcxrgj7341931b1a3ylwh5wc29-parachute-20211020-git.drv» ]
```

Gonna try some more complex examples, and try to figure out a way to have a similiar thing for `sbclPackages` itself - maybe useful for ad-hoc dependencies on slashy systems.

**Comment by Uthar** [comment]
Did the same for `sbclPackages` - this works well and makes `simple-date_slash_postgres-glue` build. But doesn't work when "merging" introduces a circular dependency. This prevents `s-sql_slash_tests` from building because `s-sql/tests`, which depends on `postmodern`, gets merged into `s-sql` while `postmodern` depends on `s-sql`. Currently trying to figure that out.

**Comment by 7c6f434c** [comment]
Well, I'd guess that losing some `/tests` systems is tolerable. I guess small circles of dependencies could be just smashed into a single package with aliases named after the parts?

**Comment by Uthar** [comment]
Yeah I was thinking of `s-sql__and__postmodern` or something like that. But wondering how this will work if something  else, then, depends on such a frankenstein, which in turn introduces another circular dependency with something else.

**Comment by 7c6f434c** [comment]
Well, if an amalgamation can be built, depending on it is safe

**Comment by Uthar** [comment]
Yeah, sounds doable
Ill write some code and see

**Comment by Uthar** [comment]
Added fixing of circular dependencies by building derivations with conflicting systems removed from the dependency tree and instead with their combined systems and asds.

For example:

(`dbi/test` has circular dependencies with `dbi-test`, `dbd-mysql`, `dbd-postgres` and `dbd-sqlite3`)
(`crane` depends on `dbi`)

```
nix-repl> sbclPackages.crane
«derivation /nix/store/063lcqwbizzj5fnsk6qx8j2m8cw98qs1-crane-20160208-git.drv»

nix-repl> builtins.elemAt sbclPackages.crane.lispLibs 4
«derivation /nix/store/gar9fya74pqfp112mj8l7i70wfbhk79d-dbi-20211020-git.drv»

nix-repl> sbclPackages.dbi
«derivation /nix/store/gar9fya74pqfp112mj8l7i70wfbhk79d-dbi-20211020-git.drv»

nix-repl> sbclPackages.dbi.systems
[ "dbi" ]

nix-repl> sbclPackages.dbi_slash_test
«derivation /nix/store/l4paxq4vl61lrbpvwbk9qkyqabp4zdmi-dbi-20211020-git.drv»

nix-repl> sbclPackages.dbi_slash_test.systems
[ "dbi" "dbi/test" "dbi-test" "dbd-sqlite3" "dbd-postgres" "dbd-mysql" ]

nix-repl> crane-and-dbi-test = sbclWithPackages (ps: [ sbclPackages.dbi_slash_test sbclPackages.crane ])
«derivation /nix/store/jf46jj8ny0bma71x67c4bc0asp0zfi4w-sbcl-with-packages.drv»

nix-repl> crane-and-dbi-test.lispLibs
[ «derivation /nix/store/l4paxq4vl61lrbpvwbk9qkyqabp4zdmi-dbi-20211020-git.drv» «derivation /nix/store/qxxqf5wc1s4brsjbvp0jwnwsz0fjnllc-crane-20160208-git.drv» ]

nix-repl> crane = builtins.elemAt crane-and-dbi-test.lispLibs 1
«derivation /nix/store/fd3xx045kja69kb68bdnfgg7g84ch4z1-crane-20160208-git.drv»

nix-repl> builtins.elemAt crane.lispLibs 4
«derivation /nix/store/l4paxq4vl61lrbpvwbk9qkyqabp4zdmi-dbi-20211020-git.drv»
# Same as sbclPackages.dbi_slash_test
```

As far as I could figure out there are only 58 such circular dependencies in QL. These are systems like `foo/test` or `foo/docs`, except for `hu.dwim.util/{authorization,finite-state-machine,flexml,production,soap,standard-process,temporary-files,threads,worker-group}` and `hu.dwim.def/namespace`.

I am working with limited data, though, because systems.txt doesn't capture the fact that something depends on a slashy system.

**Comment by Uthar** [comment]
Still thinking about this BTW.

There is a performance problem with `cl-generic` and `cl-ana` - they have big dependency trees, and Nix evaluator is taking a long to eval it. Maybe it's not caching already evaled nodes?

Also added some [CI](https://github.com/Uthar/nix-cl/actions/workflows/build-packages.yml) to test building with different CL implementations. Currently I'm aiming to build what's in `pkgs.lispPackages` because running the whole Quicklisp would be better done on a faster machine.

Trying to figure out how to put the packages into SQLite instead of making a huge nix file. This way different backends (Quicklisp, Ultralisp, cl2nix etc.) can be combined more easily. Nix can consume such a database with `builtins.fromJSON` + `sqlite3 -json pkgs.db "select ..."`. 

There are some advantages of doing it this way: 

- It's language agnostic (or more like SQL as a common ground, most languages have SQL generator libraries already). 

- There is the possibility of manually adding packages (maybe through a web form), which would be safer/easier out of the box because of ACID. So no need to write a lot of validation code. Kind of weird idea but I like it. Probably it would take some time to come up with a good schema for this.

And I was thinking, with flakes the direction is more towards decentralized/independent... well... flakes... instead of putting everything in the nixpkgs monorepo. So is there a point in putting new code in `nixpkgs#legacyPackages`?

**Comment by 7c6f434c** [comment]
About SQLite, that seems to be (from Nixpkgs point of view) more of a way to generate (possibly the same) JSON, not something happenning during the evaluation, right?

We have already seen how Nixpkgs/NixOS repo split went, and I am not any more optimistic about flakes. Given that flakes RFC has been withdrawn, I think packaging _for_ Nixpkgs can be done with a «let's wait until a flakes RFC actually passes» approach.

Of course I am not saying anything about out-of-Nixpkgs repository for faster updates of Lisp packages…

**Comment by bhankas** [comment]
There are things in current flake implementation that I like and things I'd prefer were different. But to be honest, at this point, I don't believe actual RFC is going to matter. People have started building infrastructure on top of flakes, Nix itself is getting married to its current implementation (nix-command feature) and for better or worse, I do not have a hope of any major changes in flakes going forward.

Unless you are expecting something specific and (relatively) small part of flakes to be put in cement, I don't see a point in waiting for flake RFC.

**Comment by Uthar** [comment]
> About SQLite, that seems to be (from Nixpkgs point of view) more of a way to generate (possibly the same) JSON, not something happenning during the evaluation, right?

Well I'm thinking something like:

```nix
let 
  packagesDB = fetchurl { url = "https://quicklisp.org/packages.sqlite"; hash = "..."; };
  packagesJSON = pkgs.runCommand "packages.json" {} ''
    ${pkgs.sqlite}/bin/sqlite3 -readonly -json ${packagesDB} "select * from packages" > $out
  '';
in builtins.fromJSON (builtins.readFile "${packagesJSON}")
```

Again, advantage would be having ACID + SQLite performance to help maintain the large package database, rather than editing JSON by itself (which can easily be generated almost instantly - `select * from system_view` takes 5ms on this [packages.sqlite](https://galkowski.xyz/packages.sqlite) for 5k packages)

This could be skipped and let's just distribute the json file directly, but it's another manual step... Not sure.

**Comment by 7c6f434c** [comment]
> Unless you are expecting something specific and (relatively) small part of flakes to be put in cement, I don't see a point in waiting for flake RFC.

I think some minor improvements are likely to happen from the RFC (or little changes and Nixpkgs stays as it is with very limited flake _reliance_ even if with full flake interface _provided_) — just because of how quickly things get tense specifically about ignoring the feedback every time, and the discussions seems to be focused on the parts that are indeed most questionable.

My real motivation is: I _expect_ monorepo separation to go badly, so I want to defer any effort until the enthusiasts collect all the problems.

> This could be skipped and let's just distribute the json file directly, but it's another manual step... Not sure.

Well, not even distribute, commit…

I do not know where import-from-derivation that you use here will go (in terms of Nixpkgs/Hydra), so I think in terms of it not being allowed in Nixpkgs proper.

**Comment by nagy** [comment]
> packagesDB = fetchurl { url = "https://quicklisp.org/packages.sqlite"; hash = "..."; };

Where is this sqlite file in the end supposed to be coming from? I think it is important that the history should be tracked, so that we can go back to any commit and have a working version of the code at that time. If that file (on a remote server?) is not immutable ( and versioned ), then going back in time would not be possible.

> Well, not even distribute, commit…

Yes this would be great. But keep in mind , that sqlite files are binary and diffing them is not easily done via git. There is something like this: https://gist.github.com/peteristhegreat/a028bc3b588baaea09ff67f405af2909 which converts ( and stores !) the sqlite db as a text format, that of SQL statements, into the git tree instead of the binary. This makes diffs looks nice , but has the disadvantage, that user need to setup up such a converter on their git installation in order to be able to work with it. On checkout git would then take the textual form stored in the repo and write it to a local database.

**Comment by Uthar** [comment]
> Where is this sqlite file in the end supposed to be coming from? I think it is important that the history should be tracked, so that we can go back to any commit and have a working version of the code at that time. If that file (on a remote server?) is not immutable ( and versioned ), then going back in time would not be possible.

Yes, just from some server. 
You're right that losing git history is pretty bad. Originally I thought that this file would be generated by a version-controlled script, then it would be possible to go back in time by reverting that script. But now I see than it would not be easily browsable, like a git dif would. It's also harder to debug because there's no source code to poke. Maybe let's keep the [imported.nix](https://raw.githubusercontent.com/Uthar/nix-cl/master/imported.nix) file as it is.

SHA256 hashes are stored in `packages.sqlite` between runs. It's not strictly necessary to **have** a database beforehand to make a Quicklisp import, but it just saves time. For example, when I was bumping the Quicklisp release some time ago, most of the tarball URL's haven't changed, so the import took a couple of minutes instead of a dozen.
 
 (BTW Guix also has a huge [lisp-xyz.scm](https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/lisp-xyz.scm) file for its lisp libraries. But I think it only supports SBCL and ECL)

**Comment by Uthar** [comment]
Does anyone know if it's possible to pre-configure ASDF to load some system from a different file than
 `(concatenate 'string system-name ".asd")`? I'm reading the docs but can't find anything.

**Comment by 7c6f434c** [comment]
>Does anyone know if it's possible to pre-configure ASDF to load some system from a different file than
> `(concatenate 'string system-name ".asd")`? I'm reading the docs but can't find anything.

I would assume that for that level of nuance you need to read the source and the comments.

(The comments are pretty extensive for complicated parts…)

My silly guess would be: would just load-asd on the already known file be good enough?

Of course, if there is a conflicting .asd inside the search path, we might end up with a surprise in some situations…

**Comment by Uthar** [comment]
Thanks, I should read the source. I'm just thinking how to get rid of the current mess needed to handle duplicate .asd files. If ASDF could do that, I thought that maybe something like a map of system names -> asd files would be more readable.

**Comment by Uthar** [comment]
> My silly guess would be: would just load-asd on the already known file be good enough?

We tested some time ago but it didn't work for slashy systems: https://github.com/Uthar/nix-cl/issues/1#issuecomment-1022079758

**Comment by 7c6f434c** [comment]
>Thanks, I should read the source. I'm just thinking how to get rid of the current mess needed to handle duplicate .asd files. If ASDF could do that, I thought that maybe something like a map of system names -> asd files would be more readable.

The real issue, that won't be handled _just_ by search changes, and which is the real issue in the discussion you link:

We can change _which_ .asd files get loaded, but once an .asd file get loaded, _all_ systems from the file are going to get loaded from this file. And this is probably hard to work around.

I guess we could try to do something rather silly, like mangling the unwanted system definition names in the .asd files…

**Comment by Uthar** [comment]
> We can change _which_ .asd files get loaded, but once an .asd file get loaded, _all_ systems from the file are going to get loaded from this file. And this is probably hard to work around.

Tested it a little bit, here are some results. I used:
```nix
 lispPackages_new.sbclWithPackages (p: 
   [ p.cl-async-base 
     p.cl-async-util 
     p.cl-async 
   ])
 ```
 
 as an example, because all three contain the `defsystem` forms of each other, they were just copied to create the "missing" `.asd`s.

```
* (require 'asdf)
("ASDF" "asdf" "UIOP" "uiop")

* (asdf:load-system 'cl-async-util)
T

* (asdf:system-source-directory 'cl-async-util)
#P"/nix/store/a66qzg4qgppbq1c9hwgl1hjy7gfxndc0-cl-async-util-20211020-git/"

* (asdf:load-system 'cl-async-base)
T

* (asdf:system-source-directory 'cl-async-util)
#P"/nix/store/a66qzg4qgppbq1c9hwgl1hjy7gfxndc0-cl-async-util-20211020-git/"
* (asdf:system-source-directory 'cl-async-base)
#P"/nix/store/x541kcj7sgv46vj5vr7832i76f3jz7km-cl-async-base-20211020-git/"

* (asdf:load-system 'cl-async)
T

* (asdf:system-source-directory 'cl-async)
#P"/nix/store/i386bh1i2c2aq9icd4nahi717f4dj9n9-cl-async-20211020-git/"
* (asdf:system-source-directory 'cl-async-base)
#P"/nix/store/x541kcj7sgv46vj5vr7832i76f3jz7km-cl-async-base-20211020-git/"
* (asdf:system-source-directory 'cl-async-util)
#P"/nix/store/a66qzg4qgppbq1c9hwgl1hjy7gfxndc0-cl-async-util-20211020-git/"
```

It breaks though when trying to load `cl-async` two times in a row, which is... weird:
```
* (require 'asdf)
("ASDF" "asdf" "UIOP" "uiop")
* (asdf:load-system 'cl-async)
T
* (asdf:load-system 'cl-async)
debugger invoked on a SB-INT:SIMPLE-FILE-ERROR in thread
#<THREAD "main thread" RUNNING {1004B88463}>:
  Error opening #P"/nix/store/x541kcj7sgv46vj5vr7832i76f3jz7km-cl-async-base-20211020-git/src/util/package-tmpGHU3ALSV.fasl":

    Read-only file system
0]
```
`(invoke-restart 'ASDF:CLEAR-CONFIGURATION-AND-RETRY)` makes it load, but no idea what are the consequences now.

**Comment by 7c6f434c** [comment]
>It breaks though when trying to load `cl-async` two times in a row, which is... weird:
>```
>* (require 'asdf)
>("ASDF" "asdf" "UIOP" "uiop")
>* (asdf:load-system 'cl-async)
>T
>* (asdf:load-system 'cl-async)
>debugger invoked on a SB-INT:SIMPLE-FILE-ERROR in thread
>#<THREAD "main thread" RUNNING {1004B88463}>:
>  Error opening #P"/nix/store/x541kcj7sgv46vj5vr7832i76f3jz7km-cl-async-base-20211020-git/src/util/package-tmpGHU3ALSV.fasl":
>
>    Read-only file system
>0]
>```

ASDF plan timestamps are slightly magical, and maybe cl-async does something during the load that ends up throwing off ASDF to decided the build plan is fresher than the output (even if _source files_ are older)

**Comment by teu5us** [comment]
It does, but it won't load cl-async even once unlike load-system

**Comment by 7c6f434c** [comment]
>It does, but it won't load cl-async even once unlike load-system

Hm, strange indeed…

**Comment by hraban** [comment]
I'm having problems with `cl-async`, though I can't exactly pin-point the problem because I'm no expert in neither asdf nor nix at this point.

Demo of failing build: https://github.com/hraban/nix-lisppackages-new-demo

Is this how you're supposed to use it? Specifically it's package `cl-async` that causes the problem. Other packages like `alexandria` work fine.

Full build error:

```
$ nix-build --arg pkgs "import /my/local/nixpkgs-master {}"
this derivation will be built:
  /nix/store/q5jjrwvnl6xw51dv9gzh8ppddf9qcdwy-demo-0.0.1.drv
building '/nix/store/q5jjrwvnl6xw51dv9gzh8ppddf9qcdwy-demo-0.0.1.drv'...
unpacking sources
unpacking source archive /nix/store/ij6b6n6xhllyjqlwxxa6awbzshnkjrsy-nix-lisppackages-new-demo
source root is nix-lisppackages-new-demo
patching sources
configuring
no configure script, doing nothing
building
WARNING: System definition file #P"/nix/store/snf7f23hcrzwsinb3d9zy0yknhb54f0v-cl-async-20211020-git/cl-async.asd" contains definition for system "cl-async-base". Please only define "cl-async" and secondary systems with a name starting with "cl-async/" (e.g. "cl-async/test") in that file.
WARNING: System definition file #P"/nix/store/snf7f23hcrzwsinb3d9zy0yknhb54f0v-cl-async-20211020-git/cl-async.asd" contains definition for system "cl-async-util". Please only define "cl-async" and secondary systems with a name starting with "cl-async/" (e.g. "cl-async/test") in that file.
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:SOCKET-CLOSED
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-EOF
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-INFO
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-ERROR
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-RESET
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-TIMEOUT
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-REFUSED
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-ACCEPT-ERROR
; compiling file "/nix/store/ij6b6n6xhllyjqlwxxa6awbzshnkjrsy-nix-lisppackages-new-demo/src/package.lisp" (written 01 JAN 1970 12:00:01 AM):

; wrote /private/tmp/nix-build-demo-0.0.1.drv-0/nix-lisppackages-new-demo/src/package-tmpGHU3ALSV.fasl
; compilation finished in 0:00:00.007
; compiling file "/nix/store/ij6b6n6xhllyjqlwxxa6awbzshnkjrsy-nix-lisppackages-new-demo/src/demo.lisp" (written 01 JAN 1970 12:00:01 AM):

; wrote /private/tmp/nix-build-demo-0.0.1.drv-0/nix-lisppackages-new-demo/src/demo-tmpAAURSO1.fasl
; compilation finished in 0:00:00.001
WARNING: System definition file #P"/nix/store/snf7f23hcrzwsinb3d9zy0yknhb54f0v-cl-async-20211020-git/cl-async.asd" contains definition for system "cl-async-base". Please only define "cl-async" and secondary systems with a name starting with "cl-async/" (e.g. "cl-async/test") in that file.
WARNING: System definition file #P"/nix/store/snf7f23hcrzwsinb3d9zy0yknhb54f0v-cl-async-20211020-git/cl-async.asd" contains definition for system "cl-async-util". Please only define "cl-async" and secondary systems with a name starting with "cl-async/" (e.g. "cl-async/test") in that file.
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:SOCKET-CLOSED
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-EOF
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-INFO
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-ERROR
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-RESET
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-TIMEOUT
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-REFUSED
WARNING: redefining DEFTYPE type to be a class: CL-ASYNC:TCP-ACCEPT-ERROR
WARNING: System definition file #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/cl-async-base.asd" contains definition for system "cl-async-util". Please only define "cl-async-base" and secondary systems with a name starting with "cl-async-base/" (e.g. "cl-async-base/test") in that file.
WARNING: System definition file #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/cl-async-base.asd" contains definition for system "cl-async". Please only define "cl-async-base" and secondary systems with a name starting with "cl-async-base/" (e.g. "cl-async-base/test") in that file.
Unhandled SB-INT:SIMPLE-FILE-ERROR in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                                {10010800A3}>:
  Error opening #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/package-tmpGHU3ALSV.fasl":

    Permission denied

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {10010800A3}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-INT:SIMPLE-FILE-ERROR "Error opening ~S" {1001D10CE3}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-INT:SIMPLE-FILE-ERROR "Error opening ~S" {1001D10CE3}>)
2: (INVOKE-DEBUGGER #<SB-INT:SIMPLE-FILE-ERROR "Error opening ~S" {1001D10CE3}>)
3: (ERROR SB-INT:SIMPLE-FILE-ERROR :PATHNAME #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/package-tmpGHU3ALSV.fasl" :MESSAGE "Permission denied" :FORMAT-CONTROL "Error opening ~S" :FORMAT-ARGUMENTS (#P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/package-tmpGHU3ALSV.fasl"))
4: (SB-IMPL::FILE-PERROR #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/package-tmpGHU3ALSV.fasl" 13 "Error opening ~S" #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/package-tmpGHU3ALSV.fasl")
5: (SB-IMPL::%OPEN-ERROR #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/package-tmpGHU3ALSV.fasl" 13 NIL :CREATE)
6: (OPEN #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/package-tmpGHU3ALSV.fasl" :DIRECTION :IO :ELEMENT-TYPE :DEFAULT :IF-EXISTS NIL :IF-DOES-NOT-EXIST :CREATE :EXTERNAL-FORMAT :UTF-8 :CLASS SB-SYS:FD-STREAM)
7: (UIOP/STREAM:CALL-WITH-TEMPORARY-FILE #<FUNCTION (FLET "BEFORE234" :IN UIOP/STREAM::GET-TEMPORARY-FILE) {534FA74B}> :WANT-STREAM-P NIL :WANT-PATHNAME-P T :DIRECTION :IO :KEEP T :AFTER NIL :DIRECTORY #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/" :TYPE "fasl" :PREFIX "package-tmp" :SUFFIX NIL :ELEMENT-TYPE NIL :EXTERNAL-FORMAT NIL)
8: (UIOP/LISP-BUILD:COMPILE-FILE* #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/package.lisp" :OUTPUT-FILE #P"/nix/store/h20fjqggj7mlzjd9rnbq3rvnszqfd8a8-cl-async-base-20211020-git/src/util/package.fasl" :EXTERNAL-FORMAT :UTF-8 :WARNINGS-FILE NIL)
9: (ASDF/LISP-ACTION:PERFORM-LISP-COMPILATION #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/LISP-ACTION:CL-SOURCE-FILE "cl-async-util" "src/util/package">)
10: ((SB-PCL::EMF ASDF/ACTION:PERFORM) #<unused argument> #<unused argument> #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/LISP-ACTION:CL-SOURCE-FILE "cl-async-util" "src/util/package">)
11: ((LAMBDA NIL :IN ASDF/ACTION:CALL-WHILE-VISITING-ACTION))
12: ((:METHOD ASDF/ACTION:PERFORM-WITH-RESTARTS :AROUND (T T)) #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/LISP-ACTION:CL-SOURCE-FILE "cl-async-util" "src/util/package">) [fast-method]
13: ((:METHOD ASDF/PLAN:PERFORM-PLAN (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {1004F6E873}>) [fast-method]
14: ((FLET SB-C::WITH-IT :IN SB-C::%WITH-COMPILATION-UNIT))
15: ((:METHOD ASDF/PLAN:PERFORM-PLAN :AROUND (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {1004F6E873}>) [fast-method]
16: ((:METHOD ASDF/OPERATE:OPERATE (ASDF/OPERATION:OPERATION ASDF/COMPONENT:COMPONENT)) #<ASDF/OPERATE:BUILD-OP > #<ASDF/SYSTEM:SYSTEM "demo"> :PLAN-CLASS NIL :PLAN-OPTIONS NIL) [fast-method]
17: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> #<ASDF/OPERATE:BUILD-OP > #<ASDF/SYSTEM:SYSTEM "demo">)
18: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
19: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) #<ASDF/OPERATE:BUILD-OP > #<ASDF/SYSTEM:SYSTEM "demo">) [fast-method]
20: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> ASDF/OPERATE:BUILD-OP "demo")
21: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
22: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/OPERATE:BUILD-OP "demo") [fast-method]
23: (ASDF/SESSION:CALL-WITH-ASDF-SESSION #<FUNCTION (LAMBDA NIL :IN ASDF/OPERATE:OPERATE) {1004F6528B}> :OVERRIDE T :KEY NIL :OVERRIDE-CACHE T :OVERRIDE-FORCING NIL)
24: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
25: (ASDF/SESSION:CALL-WITH-ASDF-SESSION #<FUNCTION (LAMBDA NIL :IN ASDF/OPERATE:OPERATE) {1004EE5AEB}> :OVERRIDE NIL :KEY NIL :OVERRIDE-CACHE NIL :OVERRIDE-FORCING NIL)
26: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/OPERATE:BUILD-OP "demo") [fast-method]
27: (ASDF/OPERATE:MAKE "demo")
28: (SB-INT:SIMPLE-EVAL-IN-LEXENV (ASDF/OPERATE:MAKE "demo") #<NULL-LEXENV>)
29: (EVAL-TLF (ASDF/OPERATE:MAKE "demo") 4 NIL)
30: ((LABELS SB-FASL::EVAL-FORM :IN SB-INT:LOAD-AS-SOURCE) (ASDF/OPERATE:MAKE "demo") 4)
31: ((LAMBDA (SB-KERNEL:FORM &KEY :CURRENT-INDEX &ALLOW-OTHER-KEYS) :IN SB-INT:LOAD-AS-SOURCE) (ASDF/OPERATE:MAKE "demo") :CURRENT-INDEX 4)
32: (SB-C::%DO-FORMS-FROM-INFO #<FUNCTION (LAMBDA (SB-KERNEL:FORM &KEY :CURRENT-INDEX &ALLOW-OTHER-KEYS) :IN SB-INT:LOAD-AS-SOURCE) {100105664B}> #<SB-C::SOURCE-INFO {1001056613}> SB-C::INPUT-ERROR-IN-LOAD)
33: (SB-INT:LOAD-AS-SOURCE #<SB-SYS:FD-STREAM for "file /private/tmp/nix-build-demo-0.0.1.drv-0/nix-lisppackages-new-demo/build.lisp" {1001050D03}> :VERBOSE NIL :PRINT NIL :CONTEXT "loading")
34: ((LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) #<SB-SYS:FD-STREAM for "file /private/tmp/nix-build-demo-0.0.1.drv-0/nix-lisppackages-new-demo/build.lisp" {1001050D03}> NIL)
35: (SB-FASL::CALL-WITH-LOAD-BINDINGS #<FUNCTION (LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) {618782B}> #<SB-SYS:FD-STREAM for "file /private/tmp/nix-build-demo-0.0.1.drv-0/nix-lisppackages-new-demo/build.lisp" {1001050D03}> NIL #<SB-SYS:FD-STREAM for "file /private/tmp/nix-build-demo-0.0.1.drv-0/nix-lisppackages-new-demo/build.lisp" {1001050D03}>)
36: (LOAD #<SB-SYS:FD-STREAM for "file /private/tmp/nix-build-demo-0.0.1.drv-0/nix-lisppackages-new-demo/build.lisp" {1001050D03}> :VERBOSE NIL :PRINT NIL :IF-DOES-NOT-EXIST :ERROR :EXTERNAL-FORMAT :DEFAULT)
37: ((FLET SB-IMPL::LOAD-SCRIPT :IN SB-IMPL::PROCESS-SCRIPT) #<SB-SYS:FD-STREAM for "file /private/tmp/nix-build-demo-0.0.1.drv-0/nix-lisppackages-new-demo/build.lisp" {1001050D03}>)
38: ((FLET SB-UNIX::BODY :IN SB-IMPL::PROCESS-SCRIPT))
39: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN SB-IMPL::PROCESS-SCRIPT))
40: (SB-IMPL::PROCESS-SCRIPT "build.lisp")
41: (SB-IMPL::TOPLEVEL-INIT)
42: ((FLET SB-UNIX::BODY :IN SB-IMPL::START-LISP))
43: ((FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-IMPL::START-LISP))
44: (SB-IMPL::%START-LISP)

unhandled condition in --disable-debugger mode, quitting
;
; compilation unit aborted
;   caught 1 fatal ERROR condition
error: builder for '/nix/store/q5jjrwvnl6xw51dv9gzh8ppddf9qcdwy-demo-0.0.1.drv' failed with exit code 1;
       last 10 log lines:
       > 40: (SB-IMPL::PROCESS-SCRIPT "build.lisp")
       > 41: (SB-IMPL::TOPLEVEL-INIT)
       > 42: ((FLET SB-UNIX::BODY :IN SB-IMPL::START-LISP))
       > 43: ((FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-IMPL::START-LISP))
       > 44: (SB-IMPL::%START-LISP)
       >
       > unhandled condition in --disable-debugger mode, quitting
       > ;
       > ; compilation unit aborted
       > ;   caught 1 fatal ERROR condition
       For full logs, run 'nix log /nix/store/q5jjrwvnl6xw51dv9gzh8ppddf9qcdwy-demo-0.0.1.drv'.
```

My system:

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 21.6.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.11.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

In my local nixpkgs-master checkout, I have cherry-picked two extra commits to make sbcl work:

- 5c12e0efb6108b91bce711337b0ae78b74b8dc1c
- 9a9b1952fe5d9095f3d13f0ad9e66b448f2ee91f

**Comment by hraban** [comment]
Update: I think I fixed it. Could anyone interested, please have a look and tell me if this is sane? https://github.com/NixOS/nixpkgs/pull/196818

**Comment by hraban** [comment]
@Uthar looking at imported.nix, it seems individual packages are split up into their constituent systems, each available as top-level packages in nix. E.g. `cl-async-util` and `cl-async-base`, which are basically local to `cl-async`, are both exposed alongside it. Do you know why this is? I can't import them via e.g. quicklisp; they show up in `(ql:system-apropos "cl-async")` but if you try to `(ql:quickload "cl-async-util")` you get an error.

It seems to be these packages in particular that cause grief, so I was wondering what the rationale was for making them all top-level, nix managed dependencies.

**Comment by Uthar** [comment]
Hi @hraban
I'm cooking some rice now, I will get back to you here

**Comment by Uthar** [comment]
Try like this

```patch
--- a/default.nix
+++ b/default.nix
@@ -13,9 +13,4 @@ let
     lisp = "${pkgs.sbcl}/bin/sbcl --script";
     dontStrip = true;
   };
-in
-asdf-only.overrideAttrs (final: prev: {
-  buildPhase = prev.buildPhase + ''
-    ${prev.lisp} build.lisp
-  '';
-})
+in asdf-only

```

**Comment by Uthar** [comment]
> E.g. `cl-async-util` and `cl-async-base`, which are basically local to `cl-async`, are both exposed alongside it. Do you know why this is?

I made systems exist in their own `.asd` file to ensure that there is always a single `<package>.asd` file in the asdf source registry. otherwise it's ambiguous from which one to load <package>.  

I think that asdf started warning about this some time ago, which caused a huge drama

**Comment by Uthar** [comment]
in particular to work around this problem:

>Let's say a project contains foo.asd and bar.asd. One packages foo as a nix package and places it into the immutable /nix/store. Now loading foo works fine. Later on, one might want to package bar as well. One puts both store paths in the source registry. Now ASDF will, depending on which system appears first in CL_SOURCE_REGISTRY, either try to load foo from ${bar}/foo.asd or from ${foo}/foo.asd, and, load bar from either ${bar}/bar.asd or ${foo}/bar.asd. In the ${bar} case, foo fails and bar succeeds and In the ${foo} case, foo succeeds and bar fails. This is because ${foo} only contains the fasls of foo, and vice versa. Now loading triggers a write attempt to /nix/store to compile the missing fasls.

**Comment by hraban** [comment]
cl-async-base and cl-async-util are both defined in cl-async.asd, not in their own .asd files. They're "private" packages, internal to cl-async, and shouldn't really be exposed as top-level. See the discussion in #196818 .

**Comment by Uthar** [comment]
yeah that could be better because of less source duplication, but it was easier to write quicklisp importing for me like this

**Comment by Uthar** [comment]
because of less special cases

**Comment by Uthar** [comment]
What do you think if it was possible to use Lisp packages like this?

`sbcl.pkgs.alexandria`
`sbcl.withPackages (ps: [ ps.alexandria ])`

**Comment by Uthar** [comment]
I'm also thinking about something like this in `all-packages.nix`

```
  lispImplementations = callPackage ../development/lisp-modules-new {};
  inherit (lispImplementations) abcl ccl clisp clasp ecl sbcl;
```

let me know of any thoughts

**Comment by Uthar** [comment]
Now that I think about the slashy systems problem, if there is a way to `asdf:load-system` *without* compiling the `.lisp` files, that would maybe solve it? Do you know of such a way?

**Comment by 7c6f434c** [comment]
What `asdf:load-system` even _is_ if not something to compile and load every source file?

**Comment by Uthar** [comment]
It could just load the files, there could be `asdf:load-and-compile-system` to both load and compile

**Comment by 7c6f434c** [comment]
I wonder if the most proper solution available would be just to provide ASDF a function for compiled output location that tries next to the source file, in the current build output, and in $HOME-stored cache in that order…

**Comment by Uthar** [comment]
Recently I read the python infrastructure in nixpkgs. Now what I'm thinking to do is to make secondary systems optional via overlays. So that we can still keep the fasls in /nix/store.

```nix
{

# Instead of:
sbcl-with-foo-and-bar = sbcl.withPackages (ps: [ ps.foo ps.bar_slash_baz ]);

# Use overlays:
sbcl' = sbcl.override {
  packageOverlays = self: super: {
    bar = super.bar.overrideLispAttrs (old: {
      systems = old.systems ++ [ "bar/baz" ];
    });
  };
};
sbcl-with-foo-and-bar = sbcl'.withPackages (ps: [ ps.foo ps.bar ]);

}
```

**Comment by Uthar** [comment]
see https://github.com/NixOS/nixpkgs/pull/218870
